### ISSUE TITLE
idtoken serverless sample did not work when running in Cloud Functions

### ISSUE THREAD

**Issue by VeryCrazyDog** [googleapis/google-auth-library-nodejs#1208]
I have been following the [Cloud Functions documentation](https://cloud.google.com/functions/docs/securing/authenticating#generating_tokens_programmatically) to invoke a cloud function from another cloud function.

The target URL I used is in the following format:
```
https://GCP_REGION-PROJECT_ID.cloudfunctions.net/my-function
```

Based on the sample code:
https://github.com/googleapis/google-auth-library-nodejs/blob/dd2d07a01daf9d7e766189b889fb36a9f6fb8e73/samples/idtokens-serverless.js#L36-L41

The value of the variables will be:
- `url`: `https://GCP_REGION-PROJECT_ID.cloudfunctions.net/my-function`
- `targetAudience`: `https://GCP_REGION-PROJECT_ID.cloudfunctions.net`

When calling the another cloud function, `GaxiosError` will be returned with HTTP 401 Unauthorized with error message *The access token could not be verified*.

After comparing with the JWT generated by metadata server, I found that the value of `targetAudience` should be set equal to the value of `url`. After doing so, I can call the another cloud function without error.

In conclusion, the sample should have the `targetAudience` set equal to `url`. However, I am not sure if the same applies if the target is Cloud Run.

**Comment by bcoe** [comment]
@VeryCrazyDog thanks for digging into this, and figuring out how to make things work in Cloud Functions.

Mind providing an example of the code that worked for Cloud Functions, perhaps we can add a blurb specifically for Cloud Functions.

**Comment by VeryCrazyDog** [comment]
Below is the tested code that worked on Cloud Functions. Make sure the [permission is properly granted to the caller service account](https://cloud.google.com/functions/docs/securing/authenticating#authenticating_function_to_function_calls). After deploy, invoke the function on *Google Cloud Platform Console* via the *TESTING* page with triggering event payload `{"data": {}}`.

```js
'use strict'

const functions = require('firebase-functions')
const { GoogleAuth } = require('google-auth-library')

// Hardcoded configurations
const CALLER_SERVICE_ACCOUNT = 'CALLER_SERVICE_ACCOUNT_NAME@CALLER_PROJECT_ID.iam.gserviceaccount.com'
const TARGET_URL = 'https://GCP_REGION-TARGET_PROJECT_ID.cloudfunctions.net/FUNCTION_NAME'

// Dependencies initialization
const auth = new GoogleAuth()

// Cloud function implementation
exports.idtokensServerless = functions.runWith({
  serviceAccount: CALLER_SERVICE_ACCOUNT
}).https.onCall(async (data, context) => {
  const url = TARGET_URL
  const targetAudience = TARGET_URL
  console.info(`Request ${url} with target audience ${targetAudience}`)
  const client = await auth.getIdTokenClient(targetAudience)
  const resp = await client.request({ url })
  console.info(resp.data)
})
```

**Comment by chingor13** [comment]
The sample was updated in #1218 to no longer use `.origin()`. Using the full URL for the audience works on both GCF and Cloud Run.

**Comment by guibirow** [comment]
I have opened [another issue](https://github.com/googleapis/google-auth-library-nodejs/issues/1339) to cover this, because this is a common issue created by the documentation asking people to use the serverless example provided in this repo.

If the example is called serverless, it should provide examples for cloud run and cloud functions explicitly, otherwise people will have this problem, that quite frankly, is too often.

Also, another issue people confuses, is that the audience should be the url (like mentioned by VeryCrazyDong) but it is not quite right.
The audience need to match the trigger in the cloud functions, in some cases people uses query strings or subpaths to represent REST operations, and using the url will also cause authentication issues.

Audience Examples:
valid: `https://GCP_REGION-TARGET_PROJECT_ID.cloudfunctions.net/FUNCTION_NAME`
invalid: `https://GCP_REGION-TARGET_PROJECT_ID.cloudfunctions.net/FUNCTION_NAME?query=string`
invalid: `https://GCP_REGION-TARGET_PROJECT_ID.cloudfunctions.net/FUNCTION_NAME/resource/id`
